File: exercises/python/lab07/fork_demo.py
"""Lab 07 - Process creation using os.fork(), os.getpid(), os.getppid(), os.wait()"""
import os
import sys
import time

def main():
    print("=== fork() System Call Demo ===\n")

    pid = os.fork()

    if pid < 0:
        print("fork failed", file=sys.stderr)
        sys.exit(1)
    elif pid == 0:
        # Child process
        print(f"Child  => PID: {os.getpid()}, PPID: {os.getppid()}")
        print("Child  => Doing some work...")
        time.sleep(1)
        print("Child  => Done. Exiting.")
        sys.exit(0)
    else:
        # Parent process
        print(f"Parent => PID: {os.getpid()}")
        print(f"Parent => Child PID is {pid}. Waiting...")
        child_pid, status = os.wait()
        exit_code = os.WEXITSTATUS(status)
        print(f"Parent => Child (PID {child_pid}) exited with status {exit_code}")
        print("Parent => Child process finished.")


if __name__ == "__main__":
    main()

File: exercises/cpp/lab07/fork_demo.cpp
// Lab 07 - Process Creation using fork(), getpid(), getppid(), wait(), exit()
#include <stdio.h>
#include <stdlib.h>
#include <sys/types.h>
#include <sys/wait.h>
#include <unistd.h>

int main(void) {
    printf("=== fork() System Call Demo ===\n\n");

    pid_t pid = fork();

    if (pid < 0) {
        perror("fork failed");
        return EXIT_FAILURE;
    } else if (pid == 0) {
        // Child process
        printf("Child  => PID: %d, PPID: %d\n", getpid(), getppid());
        printf("Child  => Doing some work...\n");
        sleep(1);
        printf("Child  => Done. Exiting.\n");
        exit(EXIT_SUCCESS);
    } else {
        // Parent process
        printf("Parent => PID: %d\n", getpid());
        printf("Parent => Child PID is %d. Waiting...\n", pid);
        int status;
        wait(&status);
        if (WIFEXITED(status))
            printf("Parent => Child exited with status %d\n", WEXITSTATUS(status));
        printf("Parent => Child process finished.\n");
    }

    return EXIT_SUCCESS;
}
